File: libs/innermodel-python3/innermodel_python3/innermodelprismaticjoint.py
from innermodeltransform import InnerModelTransform

class InnerModelPrismaticJoint (InnerModelTransform):
    def __init__(self, id: str, min: float, max: float, value: float, offset: float, port: int = 0,
                 axis: str = 'z', home: float = 0, parent: 'InnerModelTransform' = None):
        super (InnerModelPrismaticJoint, self).__init__ (id, 'static', 0, 0, 0, 0, 0, 0, 0, parent)
        self.value = value
        self.offset = offset
        self.min = min
        self.max = max
        self.home = home
        self.port = port
        self.axis = axis
        self.setPosition (value)

    def __repr__ (self):
        s1 = "InnerModelPrismaticJoint, id: {}, offset: {}, min: {}, max: {}, home: {}, port: {}"\
              .format (self.id, self.offset, self.min, self.max, self.home, self.port)
        s2 = ", axis: {}, pos: [{}, {}, {}], orient: [{}, {}, {}]".format (self.axis, self.tx, \
              self.ty, self.tz, self.rx, self.ry, self.rz)
        return s1 + s2

    def print (self, verbose): # redundant
        if verbose:
            print (self)

    def save (self, out, tabs: int):
        s = ""
        for _ in range (tabs):
            s += "\t"
        s += "<!-- joints cannot be saved yet -->\n"
        out.write (s)

    def getPosition (self) -> float:
        return self.value

    def setPosition (self, v: float) -> float:
        ret = None
        if v <= self.max and v >= min:
            ret = v
        else:
            if v > self.max:
                ret = self.max
            else:
                ret = self.min
        self.value = v = ret
        if self.axis is 'x':
            self.rtmat.set (0, 0, 0, v+self.offset, 0, 0)
        elif self.axis is 'y':
            self.rtmat.set (0, 0, 0, 0, v+self.offset, 0)
        elif self.axis is 'z':
            self.rtmat.set (0, 0, 0, 0, 0, v+self.offset)
        else:
            raise Exception ("internal error, no such axis %s" % self.axis)
        return ret

File: libs/innermodel/innermodelprismaticjoint.cpp
/*
 * Copyright 2016 <copyright holder> <email>
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 */

#include "innermodel/innermodelprismaticjoint.h"

InnerModelPrismaticJoint::InnerModelPrismaticJoint(QString id_, float min_, float max_, float val_, float offset_, uint32_t port_, std::string axis_, float home_, InnerModelTransform *parent_) : InnerModelTransform(id_,QString("static"),0,0,0,0,0,0, 0, parent_)
{
	#if FCL_SUPPORT==1
		collisionObject = NULL;
	#endif
	min = min_;
	max = max_;
	port = port_;
	axis = axis_;
	home = home_;
	offset = offset_;
	fixed = false;
	setPosition(val_);
}

void InnerModelPrismaticJoint::print(bool verbose)
{
	printf("Prismatic Joint: %s\n", qPrintable(id));
	if (verbose)
	{
		((QMat *)this)->print(qPrintable(id));
		getTr().print(id+"_T");
	}
}

void InnerModelPrismaticJoint::save(QTextStream &out, int tabs)
{
	for (int i=0; i<tabs; i++) out << "\t";
	out << "### joints cannot be saved yet ###\n";
}

void InnerModelPrismaticJoint::update()
{
	updateChildren();
}

float InnerModelPrismaticJoint::getPosition()
{
	return value;
}

float InnerModelPrismaticJoint::setPosition(float v)
{
	float ret;
	if (v <= max and v >= min)
	{
		ret = v;
	}
	else
	{
		if (v > max)
			ret = max;
		else
			ret = min;
	}
	value = v = ret;
	if (axis == "x")
	{
		set(0,0,0, v+offset,0,0);
	}
	else if (axis == "y")
	{
		set(0,0,0, 0,v+offset,0);
	}
	else if (axis == "z")
	{
		set(0,0,0, 0,0,v+offset);
	}
	else
	{
		QString error;
		error.sprintf("internal error, no such axis %s\n", axis.c_str());
		throw error;
	}
	return ret;
}

InnerModelNode * InnerModelPrismaticJoint::copyNode(QHash<QString, InnerModelNode *> &hash, InnerModelNode *parent)
{
	InnerModelPrismaticJoint *ret = new InnerModelPrismaticJoint(id, min, max, value, offset, port, axis, home, (InnerModelTransform *) parent);
	ret->level = level;
	ret->fixed = fixed;
	ret->children.clear();
	ret->attributes.clear();
	hash[id] = ret;
	ret->innerModel = parent->innerModel;

	for (QList<InnerModelNode*>::iterator i=children.begin(); i!=children.end(); i++)
	{
		ret->addChild((*i)->copyNode(hash, ret));
	}

	return ret;
}

